External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -185,52 +178,81 @@
         end
     else
         % Image contains only NaNs - skip this image
-    end
-
-
+	end
 end
 
-
 %% ========================================================================================
 %% ========================================================================================
 %% METHOD 3
-function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM, nEdgeVoxels)
+function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM, nEdgeVoxels, VoxelSize)
 %xASL_im_ExtrapolateLinearlyOverNaNs
 % The input of this function is 3D (iterations of other dimensions are done
 % in the main function above)
 %
 % IM          - input image
 % nEdgeVoxels - number of voxels at the edges of the non-nan values that we check for weird interpolation values, 
 % to be set to NaN and ignored in the extrapolation
+% VoxelSize   - voxel size is needed for extrapolating correctly as the transformation is given in mm and not voxels
 
 %% 0. Admin
 if nargin<2 || isempty(nEdgeVoxels)
     nEdgeVoxels = 0;
     % We can try this later, but it seems to work less well than doing the same based on the distance transform
 end
 
+if nargin < 3 || isempty(VoxelSize)
+	% The voxel-size is not provided and needs to be estimated from the transformation maps
+	% We set it here to empty in admin and estimate it later
+	VoxelSize = [];
+end
+
+%% Create masks and estimate voxel size
+% Create NaN mask for all 3D volumes together
+maskNaN = sum(isnan(IM), 4) > 0;
```

Comment:
So if we e.g. know for a border voxel where it came from in AP and LR, but not IS, then we now extrapolate it? Because then only extrapolating IS and using the information from the other directions could still be beneficial? This should never occur except for special situations where borders are smoothed so perhaps this is a good thing.

Reply:
Exactly - in theory, it should never happen that we have a Nan for AP but not for IS - cause al 3 coordinates are needed. So this should always work, or the deformation field is invalid. 

So we can resolve this?